File: KoKoslan/src/java/eval/KoKoContext.java
/**
 @author loriacarlos@gmail.com
 
*/
package kokoslan.ast;

import java.util.*;

public class KoKoContext extends HashMap<String, KoKoValue>{
	private KoKoContext parent; 
	public KoKoContext(){
		this(null);
	}
	public KoKoContext(KoKoContext parent){
		this.parent = parent;
	}
	public KoKoValue find(KoKoId id){
		KoKoValue val = get(id.getValue());
		if (val != null) return val;
		if (parent == null) throw new KoKoNotFoundId(id);
		return parent.find(id);
	}
	public void assoc(KoKoId id, KoKoValue val){
		put(id.getValue(), val);
	}
	public KoKoContext push(){
		return new KoKoContext(this);
	}
	public KoKoContext pop(){
		if ( parent == null ) throw new KoKoStackUnderflow();
		return parent;
	}
	
}

File: KoKoslan/src/kotlin/eval/KoKoContext.kt
/*
    Andrey Arguedas Espinoza
    Daniela Armas Sanchez
    Michael Chen Wang
    Kimberly Olivas Delgado
*/

package kokoslan.kt.eval

import java.util.*
import kokoslan.kt.ast.*
import kokoslan.kt.exception.*

class KoKoContext (private var parent: KoKoContext? = null) : HashMap<String, KoKoValue?>(){

	fun find(id: KoKoId): KoKoValue{
		val value: KoKoValue? = get(id.getValue())
		return value ?: return parent?.find(id) ?: throw KoKoNotFoundId(id)
	}

	fun assoc(id: KoKoId, value: KoKoValue?): Unit{
		put(id.getValue(), value)
	}

	fun push(): KoKoContext{
		return KoKoContext(this)
	}

	fun pop(): KoKoContext{
		return parent ?: throw KoKoStackUnderflow()
	}
	
}